### PR TITLE
feat: auto detect globs

### PR DESCRIPTION
**Pull request by alexzhang1030**
Now ast-grep vscode supports `globs` when binary version is above `v0.28.0`.

----

> [!IMPORTANT]
> Adds support for auto-detecting and using glob patterns in ast-grep VSCode extension when binary version is 0.28.0 or higher.
> 
>   - **Behavior**:
>     - Adds `detectSgVersion()` in `common.ts` to determine the binary version.
>     - Updates `buildCommand()` in `search.ts` to include glob patterns if supported.
>     - Modifies `IncludeFile.tsx` to update placeholder text and handle glob input based on version.
>   - **UI**:
>     - Adds `useDetectGlobFeatureAvailable()` in `useFeature.tsx` to check glob feature availability.
>     - Updates `useQuery.tsx` and `useSearch.tsx` to handle glob state in search queries.
>   - **Dependencies**:
>     - Adds `compare-versions` to `package.json` for version comparison.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ast-grep%2Fast-grep-vscode&utm_source=github&utm_medium=referral)<sup> for 09130bf24d56e39ff742454a2eae00bb53d43112. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/extension/common.ts [hunk 1]
```diff
@@ -96,3 +112,11 @@ export function streamedPromise<T>(
     }),
   )
 }
+
+export async function detectSgVersion() {
+  const semverRegex = /(\d+\.\d+\.\d+)/
+  const command = resolveBinary()
+  const raw = await execSgCommand(command, ['--version'])
+  const version = semverRegex.exec(raw)?.[0]!
```
[the remark this answers is not in the document]
Comment (alexzhang1030) [reply src/extension/common.ts:131]: I'm sure there's a value here

File: src/webview/hooks/useFeature.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { useEffect, useState } from 'react'
+import { childPort } from '../postMessage'
+import { compare } from 'compare-versions'
+
+export function useSgVersion() {
+  const [version, setVersion] = useState<string | null>(null)
+  useEffect(() => {
+    if (version) return
+    childPort.postMessage('getSgVersion')
+    childPort.onMessage('getSgVersion', setVersion)
```
[the remark this answers is not in the document]
Comment (alexzhang1030) [reply src/webview/hooks/useFeature.tsx:11]: `onMessage` does not provide any cleanup function.

**Comment by alexzhang1030** [comment]
Headless test is failed, anything I missed? @HerringtonDarkholme

File: package.json [hunk 3]
```diff
@@ -194,6 +194,8 @@
     "tsx": "4.19.0",
     "typescript": "5.5.4",
     "unport": "0.6.0",
-    "vscode-languageclient": "9.0.1"
-  }
+    "vscode-languageclient": "9.0.1",
+    "compare-versions": "^6.1.1"
+  },
+  "packageManager": "pnpm@9.11.0+sha512.0a203ffaed5a3f63242cd064c8fb5892366c103e328079318f78062f24ea8c9d50bc6a47aa3567cabefd824d170e78fa2745ed1f16b132e16436146b7688f19b"
```
Comment (HerringtonDarkholme) [inline package.json:200]: is this expected?

File: src/extension/common.ts [hunk 4]
```diff
@@ -3,9 +3,15 @@ import { Unport } from 'unport'
 import { workspace } from 'vscode'
 import type { ChildProcessWithoutNullStreams } from 'node:child_process'
 import { execFile } from 'node:child_process'
+import { compare } from 'compare-versions'
```
Comment (HerringtonDarkholme) [inline src/extension/common.ts:6]: I don't think we need another library for feature detection. When having errors, just prompting users to download the latest version is fine.

File: src/extension/common.ts [hunk 5]
```diff
@@ -39,24 +45,45 @@ export function resolveBinary() {
   return config
 }
 
-export async function testBinaryExist(command: string) {
+export async function detectUpcomingFeature() {
```
Comment (HerringtonDarkholme) [inline src/extension/common.ts:48]: ditto, I have no interest in maintaining version detection.

File: package.json (hunk 3 above)
Comment (alexzhang1030) [reply package.json]: That's weird, corepack will set packageManager to package.json if there is no packageManager filed.

**Comment by alexzhang1030** [comment]
I have rewritten the logic, now there is no auto-detect, we will force enable globs if the `include path` has any `*` char.

File: src/extension/search.ts [hunk 6]
```diff
@@ -127,7 +127,13 @@ export function buildCommand(query: SearchQuery) {
   if (query.lang) {
     args.push('--lang', query.lang)
   }
-  args.push(...includeFile.split(',').filter(Boolean))
+  const validIncludeFile = includeFile.split(',').filter(Boolean)
+  const hasGlobPattern = validIncludeFile.some(i => i.includes('*'))
+  if (hasGlobPattern) {
+    args.push(...validIncludeFile.map(i => `--globs=${i}`))
```
Comment (HerringtonDarkholme) [inline src/extension/search.ts:133]: looks nice

File: src/webview/SearchSidebar/SearchProviderMessage/index.tsx [hunk 7]
```diff
@@ -71,6 +71,13 @@ function Empty({ query }: { query: SearchQuery }) {
             .
           </li>
         ) : null}
+        {query.includeFile.length ? (
+          <li>
+            If you're using glob patterns in the include file, please ensure
+            that your version of ast-grep is above
+            <code style={codeStyle}>v0.28.0</code>.
```
Comment (HerringtonDarkholme) [inline src/webview/SearchSidebar/SearchProviderMessage/index.tsx:78]: this is good enough. thanks

**Review by HerringtonDarkholme — APPROVED** [review]

**Pull request merged**